storage/repository/repository.go

```diff
@@ -0,0 +1,198 @@
+package repository
+
+import (
+	"fmt"
+
+	"fyne.io/fyne"
+)
+
+// Repository represents a storage repository, which is a set of methods which
+// implement specific functions on a URI. Repositories are registered to handle
+// specific URI schemes, and the higher-level functions that operate on URIs
+// internally look up an appropriate method from the relevant Repository.
+//
+// The repository interface includes only methods which must be implemented at
+// a minimum. Without implementing all of the methods in this interface, a URI
+// would not be usable in a useful way. Additional functionality can be exposed
+// by using interfaces which extend Repository.
+//
+// Repositories are registered to handle a specific URI scheme (or schemes)
+// using the Register() method. When a higher-level URI function such as
+// storage.Copy() is called, the storage package will internally look up
+// the repository associated with the scheme of the URI, then it will use
+// a type assertion to check if the repository implements CopyableRepository.
+// If so, the Copy() function will be run from the repository, otherwise
+// storage.Copy() will return NotSupportedError. This works similarly for
+// all other methods in repository-related interfaces.
+//
+// Note that a repository can be registered for multiple URI schemes. In such
+// cases, the repository must internally select and implement the correct
+// behavior for each URI scheme.
+//
+// NOTE: most developers who use Fyne should *not* generally attempt to
+// call repository methods directly. You should use the methods in the storage
+// package, which will automatically detect the scheme of a URI and call into
+// the appropriate repository.
+//
+// Since: 2.0.0
+type Repository interface {
```

Comment:
Some Repositories will want to provide custom implementations of the URI interface, it was previously discussed that `ParseURI(string) fyne.URI` could serve this purpose and, although I initially disagreed, I now believe it to be the right approach

Reply:
My understanding here was that we were discoraging external URI implementations.
That is why there was discussion about making it a struct. It's still an interface for compatibility only IIRC.

Reply:
I thought it was the way to go all along. I will re-add it and provide a `GenericParseURI()` function.

Reply:
I'm pretty sure this puts us back into breaking sepration of concerns territory...
What is the usecase we cannot do with a single URI implementation?

Yes I thought that implementing URIs would be the way to go - but then repository replaced the functionality so we only have 1 type of URI. If we follow this path then we have both...

Reply:
As I presented in Slack (https://gophers.slack.com/archives/CK5U4BU87/p1610463573476400):

> "For example, a file called "hello" may be stored in IPFS and get a URI of `ipfs://qmxzzpcazv6tw1tvicf9poare9kkb1fwmzbvamytdwvshe`, but a developer implementing an IPFS repository may want `uri.Name()` to return "hello" instead of "qmxzzpcazv6tw1tvicf9poare9kkb1fwmzbvamytdwvshe""

Reply:
Where was the URI returned from?

Reply:
As the URI is always sent from a repository it seems like it should be the repo responsibility to return the correct one.

Reply:
We also discussed a `Canonical(URI)` method if we found any use-cases that meant some URI translation was required.

Reply:
> Where was the URI returned from?

Currently the only way is `storage.ParseURI`.

> As the URI is always sent from a repository

This is not true.

I would not opposed to `Canonical(URI)` but it is an extra step. Adding `ParseURI(string) (URI, error)` to `Repository` feels cleaner, but I would also not be opposed to another interface that can be implemented to provide a custom `ParseURI(string) (URI error)`

Reply:
> Currently the only way is storage.ParseURI.

This doesn't seem right. The way to browse to a URI is through Listable code... The repository returns the child URI from it's internals.
Normally you would only be re-hydrating a URI after storage of it, in which case it should have come from the repository previously.

> I would also not be opposed to another interface

It certainly seems strange to have custom URI overrides provided for in the base Repository but no support for parent or child - which seem far more common.

Reply:
> another interface that can be implemented to provide a custom ParseURI(string) (URI error), tho I'm not sure what that interface would be called.

I am opposed to this, since we agreed on not having magic default, which `storage.ParseURI()` would need in such a case. I guess one could argue that `storage.ParseURI()` wouldn't be doing magic per-se, since it's returning a non-repository-specific URI instance, whereas `Repository.ParseURI()` or `ParseableRepository.ParseURI()` as the case may be could be returning a repository-specific URI type.

Reply:
Alright, it seems we came to a consensus on this in chat. Repositories that want to opt-in to providing their own URI parser can by implementing `CanonicalRepository`, something like:

```go
type CanonicalRepository interface {
  Repository
  ParseURI(string) (fyne.URI, error)
}
```

The developer still uses only one point of entry into URIs: `storage.ParseURI(string) (fyne.URI, error)`

Charles is updating the proposal.

Reply:
Resolved. Discussion documented [here](https://github.com/fyne-io/fyne/pull/1768#issuecomment-758801153). This conversation will close after the needed code changes are committed.

Reply:
101b02cc6893bbff274918f57e5c85629f89a215